### PR TITLE
Adding description of IREE Codegen pass pipeline

### PR DESCRIPTION
**Pull request by MaheshRavishankar**
Currently adding description of just the MHLO -> SPIR-V pass pipeline.

### CODE REVIEW

File: docs/codegen_passes.md [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+
+- The `scf.parallel` operations represent parallel loops that are to be
+  distributed across workgroups. The distribution here assumes that the number of workgroups along each dimension is equal to the number of iterations of the `scf.parallel`
+  operation.
+
+- Linalg operations that are not tiled, and are therefore __not within__ `scf`
+  operations, are lowered to loops. The resulting outer `scf.parallel` operations are
+  collapsed to have a single induction variable. This loop is then distributed
+  across workitems using their `GlobalInvocationId` (which is same as
+  `blockIdx.x * blockDim.x + threadIdx.x` in CUDA terminology).
```
Comment (antiagainst) [inline docs/codegen_passes.md:448]: what about `y` and `z` dimensions?

File: docs/codegen_passes.md [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+
+- Linalg operations that are tiled, and are therefore __within__ `scf` operations,
+  are lowered to loops and the iterations of the `scf.parallel` operations are
+  mapped to workitems using their `LocalInvocationId` (which is same as
+  `threadIdx.x` in CUDA terminology). Note that these operations are tagged with
```
Comment (antiagainst) [inline docs/codegen_passes.md:453]: what about `y` and `z` dimensions?

File: docs/codegen_passes.md [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+  the `workgroup` marker which makes it easy to disambiguate from the case where
+  Linalg operations are outside of `scf` operations. Here too, the distribution assumes that the workgroup size is greater than or equal to the number of
+  iterations of the partitioned loop.
+
+These transformations are applied by the [`ConvertToGPUPass`][ConvertToGPU]. Below is the result of applying this pass to Snippet 7. The outer `scf.parallel` loop is distributed across workgroups. The tiled `linalg.matmul` operation is lowered to loops, and the outer `scf.parallel` operation generated here are distributed across workitems within the workgroup.
```
Comment (antiagainst) [inline docs/codegen_passes.md:458]: Nit: Wrap to 80 characters? (Other places too.) Otherwise it might cause trouble when importing/exporting to Google3.

Comment (antiagainst) [inline docs/codegen_passes.md:458]: the second "outer `scf.parallel`" should be inner ?

File: docs/codegen_passes.md [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+
+-   A map is constructed that for a given `tensor` records the `memref` value to
+    use during the conversion. In this map the `tensor` value used in the
+    `hal.interface.store.tensor` is mapped to the `memref` value returned by the
+    created `iree.placeholder` operation.
+
+-   The Dialect Conversion framework is used to implement a set of patterns that
+    convert from operations on `tensor`s to operation on `memref`s,
+
+    -   A `hal.interface.load.tensor`, is replacd with an `iree.placeholder` to
```
Comment (stellaraccident) [inline docs/codegen_passes.md:206]: Typo: replaced

File: docs/codegen_passes.md [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+       linalg.yield %5 : f32
+     }: tensor<10x5xf32>, tensor<10x5xf32> -> tensor<10x5xf32>
+```
+<a name="snippet3"></a>
+Snippet 3 : MHLO to Linalg conversion for [element-wise operations](#snippet2)
+
+At the time of writing the representation of Linalg on `tensor`s does not model
+reduction iterator types completely. Specifically, the reduction in Linalg is
+modeled using read-modify-write approach, i.e. each iteration of the reduction
+loop reads the value stored in the the output, adds its contribution, and writes
```
Comment (hanhanW) [inline docs/codegen_passes.md:158]: s/the the/the

File: docs/codegen_passes.md [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+back to the same location. This means the output has to be *initialized* to the
+null element of the reduction operator (i.e. 0 if the reduction is done using
+addition). This works for operations on buffers. Since tensors are SSA values
+they cannot be updated in-place. As a result, the reduction semantics does not
+map as well to `tensor`s. For now it is treated as a convention that when the
+linalg operation is converted to use `memref`s it has to be initialized
```
Comment (hanhanW) [inline docs/codegen_passes.md:164]: Let's be consistent.
s/linalg/Linalg

File: docs/codegen_passes.md [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+        -   `mhlo.reduce_window`.
+
+        Since the specification of the Linalg operations require the output
+        `memref` to be initialized appropriately, a `linalg.fill` operation is
+        used to achieve this.
+
+__Roadmap Note__ : Right now the code-generation pipeline relies on fusion of
+operations on tensor level. In the near future, we want to be able to fuse
+operations like `linalg.matmul` and `linalg.conv` with consumers/producers that
+are element-wise operations using the [fusion of linalg operation on
```
Comment (hanhanW) [inline docs/codegen_passes.md:254]: s/linalg/Linalg

**Pull request merged**